### PR TITLE
Fix FSDP issues when "random" forecast policies are used

### LINKED ISSUE #1984: Random forecast policy and related policies not functional
### What happened?

Currently FSDP does not seem to work when the forecast policy is for example "random" and not "fixed".

This issue is related to pushforward trick #1904. A PR should be merged first that makes the polices functional prior to PR #1904. 

### What are the steps to reproduce the bug?

_No response_

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### PR DESCRIPTION
**Pull request by SavvasMel**
## Description

This PR introduces two changes:

1) Add a new forecasting policy called "sequential_prob". This means that a specific sequence of forecast steps can be selected with specific probabilities. For example the forecast steps below have been chosen with some spike probabilities:

```
0: 000 : 00010/00125 : 000010 : loss = 3.0734E-01 (lr=1.64E-06, fsteps=[19, 5, 5, 3, 5, 5, 5, 5, 3, 3, 2], s/sec=0.167)
0: 	
0: LossPhysical.ERA5.mse.avg : 6.1012E-01 	
0: LossPhysical.loss_avg : 3.0734E-01 	
0: 
0: 
0: 000 : 00020/00125 : 000020 : loss = 3.0750E-01 (lr=3.34E-06, fsteps=[5, 1, 1, 5, 3, 5, 5, 1, 19, 5], s/sec=0.482)
0: 	
0: LossPhysical.ERA5.mse.avg : 6.1149E-01 	
0: LossPhysical.loss_avg : 3.0750E-01 	

```

2) When the forecast policy is the one described above or when it is "random", FSDP complains because the different number of forecast steps are not aligned well among the workers. This PR fixes this issue but introducing a rank-independent random generator.

Instructions:

Make the changes in the config, something like:

```
  forecast :
      time_step: 06:00:00
      offset: 1
      num_steps: [3,10]
      policy: "random"
```

or

```
  forecast :
      time_step: 06:00:00
      offset: 1
      num_steps: [3,10]
      policy: "sequential_prob"
      fstep_probabilities: [0.5,0.5]
```

If run from `launch_slurm`:

```
training_config.forecast.num_steps=[3,10]
training_config.forecast.policy="random"

```

## Issue Number

Closes https://github.com/ecmwf/WeatherGenerator/issues/1984

## Checklist before asking for review

-   [X] I have performed a self-review of my code
-   [X] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [X] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 1]
```diff
@@ -326,19 +330,40 @@ def reset(self):
...
 
         # forecast time steps
+        # Use a rank-independent RNG so that all FSDP ranks generate the same
+        # num_forecast_steps schedule; self.rng is rank-specific (seeded from
+        # data_loader_rng_seed which includes the rank factor). The same
+        # worker_workset seed formula is used to  but exclude the rank component.
+        worker_info = torch.utils.data.get_worker_info()
+        worker_id = worker_info.id if worker_info is not None else 0
+        forecast_seed = self._base_rng_seed * ((worker_id + 1) * 37) * (self.mini_epoch + 13) * 7
+        forecast_rng = np.random.default_rng(forecast_seed)
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: Shouldn't this be a class variable that is initialized in the constructor?

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 2]
```diff
@@ -326,19 +330,40 @@ def reset(self):
             self.perms = self.rng.permutation(self.perms)
 
         # forecast time steps
+        # Use a rank-independent RNG so that all FSDP ranks generate the same
+        # num_forecast_steps schedule; self.rng is rank-specific (seeded from
+        # data_loader_rng_seed which includes the rank factor). The same
+        # worker_workset seed formula is used to  but exclude the rank component.
+        worker_info = torch.utils.data.get_worker_info()
+        worker_id = worker_info.id if worker_info is not None else 0
+        forecast_seed = self._base_rng_seed * ((worker_id + 1) * 37) * (self.mini_epoch + 13) * 7
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: We use this formula twice (also at the bottom of the file). Please put into a small function.

Comment (SavvasMel) [reply src/weathergen/datasets/multi_stream_data_sampler.py]: Done

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 1 above)
Comment (SavvasMel) [reply src/weathergen/datasets/multi_stream_data_sampler.py]: Done

**Comment by MatKbauer** [comment]
Two questions:
1. What does `fsteps=[19, 5, 5, 3, 5, 5, 5, 5, 3, 3, 2]` mean? Is this the steps used in each iteration? Not sure whether we need this logged to the console.
2. Can you elaborate on why we need a separate random number generator?

**Comment by SavvasMel** [comment]
> Two questions:
> 
> 1. What does `fsteps=[19, 5, 5, 3, 5, 5, 5, 5, 3, 3, 2]` mean? Is this the steps used in each iteration? Not sure whether we need this logged to the console.
> 2. Can you elaborate on why we need a separate random number generator?

1. I have not included the logging to the console. The above it is just illustrational. It means the forecast steps for each iteration.

2. So mainly every FSDP rank need the same seed so that for each iteration as you see above they can generate exactly the same "array" of forecast steps. So every FSDP rank now generates the same list e.g. `[5, 1, 1, 5, 3, 5, 5, 1, 19, 5]`

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 3]
```diff
@@ -268,11 +275,22 @@ def reset(self) -> tuple[Sequence[int], Sequence[int]]:
...
+        worker_id = worker_info.id if worker_info is not None else 0
         self.rng = np.random.default_rng(self.data_loader_rng_seed)
         fsm = self._get_fsm()
         self.check_samples(fsm)
         perms = self._calc_baseperms(fsm)
 
+        # Use a rank-independent RNG so that all FSDP ranks generate the same
+        # num_forecast_steps schedule; self.rng is rank-specific (seeded from
+        # data_loader_rng_seed which includes the rank factor). The same
+        # worker_workset seed formula is used to  but exclude the rank component.
```
Comment (grassesi) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: It is not the same formular as in worker_workset:
For the dataloading rng we essentially do `seed *= fixed_factor *  (current_mini_epoch+13) `
Here we only do `seed = base_seed * fixed_factor * (current_mini_epoch+13)`. I dont think it matters here since we only need the generators to be uncorrelated between the mini-epochs.

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 4]
```diff
@@ -140,7 +141,13 @@ def __init__(self, cf: Config, mode_cfg: dict, stage: Stage):
         nw = cf.data_loading.num_workers
         self.data_loader_rng_seed = rs if rs > nw else rs * 97
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py:159]: Why should we use the *worker-independent* mult factor here?

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 5]
```diff
@@ -297,12 +315,25 @@ def reset(self) -> tuple[Sequence[int], Sequence[int]]:
...
-            fs = self.rng.integers(
+            fs = self.forecast_rng.integers(
                 low=self.list_num_forecast_steps.min(),
                 high=fsm + 1,
                 size=len_dt,
                 dtype=np.int64,
             )
+
+        elif self.forecast_policy in ("sequential_prob"):
+            assert np.isclose(sum(self.fstep_probabilities), 1.0), "Probabilities must sum to 1.0"
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: `self.fstep_probabilities` can be None here

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 3 above)
Comment (clessig) [reply src/weathergen/datasets/multi_stream_data_sampler.py]: If we need a second rng here that is synchronized across workers and ranks, then it should be reflected in the name (rng_synced_workers). It might be useful for other things

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 6]
```diff
@@ -818,3 +847,6 @@ def worker_workset(self):
             )
 
         return iter_start, iter_end
+
+    def _mix_rng_seed(self, base_seed, worker_id, mini_epoch, rank_factor=1):
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: I don't think this is a good function name: it's not mixing something (what is mixed with what?) but it's desynchronizing or randomizing the seed across workers.

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 5 above)
Comment (SavvasMel) [reply src/weathergen/datasets/multi_stream_data_sampler.py]: You mean it "should also" be None, or that if it is None then I will get an assertion error?

The second is welcome to happen if someone selects `sequential_prob` as policy.

**Comment by kctezcan** [comment]
I started two small runs on Säntis with 
```
"forecast":{
      "time_step":"06:00:00",
      "num_steps":[ 3, 11],
      "offset":1,
      "policy":"random"
   }
```
and 
```
"forecast":{
      "time_step":"06:00:00",
      "num_steps":[ 3, 11  ],
      "offset":1,
      "policy":"sequential_prob",
      "fstep_probabilities":[ 0.5, 0.5  ]
   }
```
both completed the training successfully

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 7]
```diff
@@ -819,3 +848,6 @@ def worker_workset(self):
...
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: There is nothing mixed here. Can we call this function _advance_rng_seed()

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 8]
```diff
@@ -298,12 +316,25 @@ def reset(self) -> tuple[Sequence[int], Sequence[int]]:
...
-            fs = self.rng.integers(
+            fs = self.forecast_rng.integers(
                 low=self.list_num_forecast_steps.min(),
                 high=fsm + 1,
                 size=len_dt,
                 dtype=np.int64,
             )
+
+        elif self.forecast_policy in ("sequential_prob"):
+            assert np.isclose(sum(self.fstep_probabilities), 1.0), "Probabilities must sum to 1.0"
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: Check this in the constructor

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 9]
```diff
@@ -140,7 +141,13 @@ def __init__(self, cf: Config, mode_cfg: dict, stage: Stage):
         nw = cf.data_loading.num_workers
         self.data_loader_rng_seed = rs if rs > nw else rs * 97
 
+        # store base seed before rank-specific modification in worker_workset();
+        # used to create a rank-independent RNG for forecast step sampling so that
+        # all ranks agree on num_forecast_steps (required by FSDP)
+        self._base_rng_seed = self.data_loader_rng_seed
+
         self.rng = None
+        self.forecast_rng = None
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: The naming is not very clear here. Can we call this `rank_synchronized_rng` and above `worker_rng`

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 10]
```diff
@@ -269,11 +276,22 @@ def reset(self) -> tuple[Sequence[int], Sequence[int]]:
...
         self.rng = np.random.default_rng(self.data_loader_rng_seed)
         fsm = self._get_fsm()
         self.check_samples(fsm)
         perms = self._calc_baseperms(fsm)
 
+        # Use a rank-independent RNG so that all FSDP ranks generate the same
+        # num_forecast_steps schedule; self.rng is rank-specific (seeded from
+        # data_loader_rng_seed which includes the rank factor). The same
+        # worker_workset seed formula is used to  but exclude the rank component.
+
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: No empty line between comment and related code

**Review by clessig — COMMENTED** [review]
Thanks for the work. Just some minor things.

**Review by MatKbauer — APPROVED** [review]
Thanks for implementing this feature, Savvas.

**Comment by clessig** [comment]
> I started two small runs on Säntis with
> 
> ```
> "forecast":{
>       "time_step":"06:00:00",
>       "num_steps":[ 3, 11],
>       "offset":1,
>       "policy":"random"
>    }
> ```
> 
> and
> 
> ```
> "forecast":{
>       "time_step":"06:00:00",
>       "num_steps":[ 3, 11  ],
>       "offset":1,
>       "policy":"sequential_prob",
>       "fstep_probabilities":[ 0.5, 0.5  ]
>    }
> ```
> 
> both completed the training successfully

Which configs did you test? Was this with multiple streams? With student-teacher training?

**Comment by SavvasMel** [comment]
> > I started two small runs on Säntis with
> > ```
> > "forecast":{
> >       "time_step":"06:00:00",
> >       "num_steps":[ 3, 11],
> >       "offset":1,
> >       "policy":"random"
> >    }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```
> > "forecast":{
> >       "time_step":"06:00:00",
> >       "num_steps":[ 3, 11  ],
> >       "offset":1,
> >       "policy":"sequential_prob",
> >       "fstep_probabilities":[ 0.5, 0.5  ]
> >    }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > both completed the training successfully
> 
> Which configs did you test? Was this with multiple streams? With student-teacher training?

Personally, I have tested and it works with multiple streams (ERA5, synop and NNTMS). I have not tested with student-teacher training. Happy to test again if you send me some configs.

**Comment by kctezcan** [comment]
> Which configs did you test? Was this with multiple streams? With student-teacher training?

My test was with forecasting and ERA5
